src/transformers/trainer.py

```diff
@@ -230,25 +230,32 @@ def get_train_dataloader(self) -> DataLoader:
         """
         Returns the training :class:`~torch.utils.data.DataLoader`.
         """
-        if self.train_dataset is None:
-            raise ValueError("Trainer: training requires a train_dataset.")
-        if is_torch_tpu_available():
-            train_sampler = get_tpu_sampler(self.train_dataset)
+        if not isinstance(self.train_dataset, torch.utils.data.IterableDataset):
+            if self.train_dataset is None:
+                raise ValueError("Trainer: training requires a train_dataset.")
+            if is_torch_tpu_available():
+                train_sampler = get_tpu_sampler(self.train_dataset)
+            else:
+                train_sampler = (
+                    RandomSampler(self.train_dataset)
+                    if self.args.local_rank == -1
+                    else DistributedSampler(self.train_dataset)
+                )
+            data_loader = DataLoader(
+                self.train_dataset,
+                batch_size=self.args.train_batch_size,
+                sampler=train_sampler,
+                collate_fn=self.data_collator,
+                drop_last=self.args.dataloader_drop_last,
+            )
         else:
-            train_sampler = (
-                RandomSampler(self.train_dataset)
-                if self.args.local_rank == -1
-                else DistributedSampler(self.train_dataset)
+            data_loader = DataLoader(
```

Comment:
why about TPU + Iterable Dataset case?
I think the train_sampler logic needs to be unnested.
(Full disclosure I don't know this code well.)

Reply:
Yeah I agree this looks ugly. I'll update it - thanks.